File: web/src/api/consumerRequestMode.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0.
 */
import client from './client';

export type RequestMode = 'POP' | 'PULL';
export interface RequestModeSelection {
  instanceId: string;
  topic: string;
  group: string;
}
export interface BrokerRequestMode {
  brokerName: string;
  address: string;
  mode: RequestMode;
  popShareQueueNum: number;
  explicit: boolean;
  serverLoadBalancerEnable: string;
}
export interface RequestModePreview {
  topic: string;
  group: string;
  brokers: BrokerRequestMode[];
}
export interface RequestModeReceipt {
  brokers: {
    before: BrokerRequestMode;
    status: 'CONFIRMED' | 'UNCHANGED' | 'UNKNOWN' | 'NOT_ATTEMPTED';
    observed: BrokerRequestMode | null;
  }[];
}
export async function previewConsumerRequestMode(
  params: RequestModeSelection,
  signal?: AbortSignal,
) {
  const response = await client.get<{ data: RequestModePreview }>('/consumer-request-mode', {
    params,
    signal,
  });
  return response.data.data;
}
export async function applyConsumerRequestMode(
  selection: RequestModeSelection,
  mode: RequestMode,
  popShareQueueNum: number,
  expected: BrokerRequestMode[],
  signal?: AbortSignal,
) {
  const response = await client.post<{ data: RequestModeReceipt }>(
    '/consumer-request-mode',
    { ...selection, mode, popShareQueueNum, expected },
    { signal, timeout: 0 },
  );
  return response.data.data;
}

File: server/src/main/java/org/apache/rocketmq/studio/instance/group/ConsumerRequestMode.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0.
 */
package org.apache.rocketmq.studio.instance.group;

import java.util.List;

public final class ConsumerRequestMode {
    private ConsumerRequestMode() { }

    public record Broker(String brokerName, String address, String mode, int popShareQueueNum,
            boolean explicit, String serverLoadBalancerEnable) { }

    public record Preview(String topic, String group, List<Broker> brokers) { }

    public record Request(String instanceId, String topic, String group, String mode,
            int popShareQueueNum, List<Broker> expected) { }

    public record Outcome(Broker before, String status, Broker observed) { }

    public record Receipt(List<Outcome> brokers) { }
}
